File: 1302.层数最深叶子节点的和.py
#
# @lc app=leetcode.cn id=1302 lang=python3
#
# [1302] 层数最深叶子节点的和
#

# @lc code=start
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def __init__(self):
        self.maxdepth=-1
        self.total=0
    def deepestLeavesSum(self, root: Optional[TreeNode]) -> int:
        def dfs(node,depth):
            if not node:
                return
            print(depth)
            if depth>self.maxdepth:
                self.maxdepth,self.total=depth,node.val
            elif depth==self.maxdepth:
                self.total+=node.val
            dfs(node.left,depth+1)
            dfs(node.right,depth+1)
        dfs(root,0)
        return self.total
# @lc code=end

# Solution1:
# class Solution:
#     def deepestLeavesSum(self, root: Optional[TreeNode]) -> int:
#         q=Queue()
#         q.put(root)
#         while not q.empty():
#             tmpVal=0
#             for _ in range(q.qsize()):
#                 tmp=q.get()
#                 tmpVal+=tmp.val
#                 if tmp.left!=None:
#                     q.put(tmp.left)
#                 if tmp.right!=None:
#                     q.put(tmp.right)
#             if q.empty():
#                 return tmpVal


# dfs:
# Solution2:
# class Solution:
#     def __init__(self):
#         self.maxdepth=-1
#         self.total=0
#     def deepestLeavesSum(self, root: Optional[TreeNode]) -> int:
#         def dfs(node,depth):
#             if not node:
#                 return
#             print(depth)
#             if depth>self.maxdepth:
#                 self.maxdepth,self.total=depth,node.val
#             elif depth==self.maxdepth:
#                 self.total+=node.val
#             dfs(node.left,depth+1)
#             dfs(node.right,depth+1)
#         dfs(root,0)
#         return self.total

File: 1302.层数最深叶子节点的和.java
import java.util.LinkedList;
import java.util.Queue;

import javax.swing.tree.TreeNode;

/*
 * @lc app=leetcode.cn id=1302 lang=java
 *
 * [1302] 层数最深叶子节点的和
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    private int total=0;
    private int maxDep=-1;
    public int deepestLeavesSum(TreeNode root) {
        dfs(root,0);
        return total;
    }
    void dfs(TreeNode treeNode,int dep){
        if(treeNode==null){
            return;
        }
        if(dep>maxDep){
            maxDep=dep;
            total=treeNode.val;
        }else if(dep==maxDep){
            total+=treeNode.val;
        }
        dfs(treeNode.left,dep+1);
        dfs(treeNode.right, dep+1);
    }
}
// @lc code=end

// Solution1:
// class Solution {
//     public int deepestLeavesSum(TreeNode root) {
//         Queue<TreeNode> queue=new LinkedList<>();
//         int res=0;
//         queue.offer(root);
//         while(!queue.isEmpty()){
//             res=0;
//             int size=queue.size();
//             for(int i=0;i<size;i++){
//                 TreeNode treeNode=queue.poll();
//                 res+=treeNode.val;
//                 if(treeNode.left!=null){
//                     queue.offer(treeNode.left);
//                 }
//                 if(treeNode.right!=null){
//                     queue.offer(treeNode.right);
//                 }
//             }
//             System.out.println(res);
//         }
//         return res;
//     }
// }



// Solution2:
// class Solution {
//     private int total=0;
//     private int maxDep=-1;
//     public int deepestLeavesSum(TreeNode root) {
//         dfs(root,0);
//         return total;
//     }
//     void dfs(TreeNode treeNode,int dep){
//         if(treeNode==null){
//             return;
//         }
//         if(dep>maxDep){
//             maxDep=dep;
//             total=treeNode.val;
//         }else if(dep==maxDep){
//             total+=treeNode.val;
//         }
//         dfs(treeNode.left,dep+1);
//         dfs(treeNode.right, dep+1);
//     }
// }